data/transactions/logic/eval.go

```diff
@@ -4707,76 +4829,111 @@ func opItxnNext(cx *EvalContext) error {
-	if sv.argType() != StackBytes || len(sv.Bytes) != crypto.DigestSize {
-		return basics.Address{}, fmt.Errorf("not an address")
+// assignAccount is used to convert a stackValue into a 32-byte account value,
+// enforcing any "availability" restrictions in force.
+func (cx *EvalContext) assignAccount(sv stackValue) (basics.Address, error) {
+	_, err := sv.address()
+	if err != nil {
+		return basics.Address{}, err
 	}
 
 	addr, _, err := cx.accountReference(sv)
 	return addr, err
 }
 
-// availableAsset is used instead of asaReference for more recent opcodes that
-// don't need (or want!) to allow low numbers to represent the asset at that
-// index in ForeignAssets array.
-func (cx *EvalContext) availableAsset(sv stackValue) (basics.AssetIndex, error) {
+// assignAsset is used to convert a stackValue to a uint64 assetIndex, reporting
+// any errors due to availability rules or type checking.
+func (cx *EvalContext) assignAsset(sv stackValue) (basics.AssetIndex, error) {
 	uint, err := sv.uint()
 	if err != nil {
-		return basics.AssetIndex(0), err
+		return 0, err
 	}
 	aid := basics.AssetIndex(uint)
 
+	if cx.availableAsset(aid) {
+		return aid, nil
+	}
+
+	return 0, fmt.Errorf("invalid Asset reference %d", aid)
+}
+
+// availableAsset determines whether an asset is "available". Before
+// resourceSharingVersion, an asset had to be available for for asset param
+// lookups, asset holding lookups, and asset id assignments to inner
+// transactions. After resourceSharingVersion, the distinction must be more fine
+// grained. It must be available for asset param lookups, or use in an asset
+// transaction (axfer,acfg,afrz), but not for holding lookups or assignments to
+// an inner static array.
+func (cx *EvalContext) availableAsset(aid basics.AssetIndex) bool {
 	// Ensure that aid is in Foreign Assets
 	for _, assetID := range cx.txn.Txn.ForeignAssets {
 		if assetID == aid {
-			return aid, nil
+			return true
 		}
 	}
 	// or was created in group
 	if cx.version >= createdResourcesVersion {
-		for _, assetID := range cx.available.asas {
+		for _, assetID := range cx.available.createdAsas {
 			if assetID == aid {
-				return aid, nil
+				return true
 			}
 		}
 	}
 
-	return basics.AssetIndex(0), fmt.Errorf("invalid Asset reference %d", aid)
+	// or some other txn mentioned it
+	if cx.version >= resourceSharingVersion {
+		if _, ok := cx.available.sharedAsas[aid]; ok {
+			return true
+		}
+	}
+
+	return false
 }
 
-// availableApp is used instead of appReference for more recent (stateful)
-// opcodes that don't need (or want!) to allow low numbers to represent the app
-// at that index in ForeignApps array.
-func (cx *EvalContext) availableApp(sv stackValue) (basics.AppIndex, error) {
+// assignApp is used to convert a stackValue to a uint64 appIndex, reporting
+// any errors due to availability rules or type checking.
+func (cx *EvalContext) assignApp(sv stackValue) (basics.AppIndex, error) {
```

Comment:
maybe move with the other assigns

Reply:
I moved `assignAccount` instead.  My intent was to keep the `assignX` functions close to the function they are built on (in each case they are thin wrappers over the next function in the file). So the app stuff together, asset stuff together, rather than `assign`s together and `available`s together.